src/client/pythonEnvironments/collection/environmentsReducer.ts

```diff
@@ -0,0 +1,129 @@
+import {
+    ILocator, IPythonEnvsIterator, PythonEnvUpdatedEvent, QueryForEvent,
+} from '../base/locator';
+import { PythonEnvsChangedEvent } from '../base/watcher';
+
+export class PythonEnvsReducer implements ILocator {
+    public get onChanged(): Event<PythonEnvsChangedEvent> {
+        return this.pythonEnvsManager.onChanged;
+    }
+
+    constructor(private readonly pythonEnvsManager: ILocator) {}
+
+    public resolveEnv(env: string | PythonEnvInfo): Promise<PythonEnvInfo | undefined> {
+        return this.pythonEnvsManager.resolveEnv(env);
+    }
+
+    public iterEnvs(query?: QueryForEvent<PythonEnvsChangedEvent>): IPythonEnvsIterator {
+        const didUpdate = new EventEmitter<PythonEnvUpdatedEvent | null>();
+        const iterator: IPythonEnvsIterator = this.iterEnvsIterator(didUpdate, query);
+        iterator.onUpdated = didUpdate.event;
+        return iterator;
+    }
+
+    private async* iterEnvsIterator(
+        didUpdate: EventEmitter<PythonEnvUpdatedEvent | null>,
+        query?: QueryForEvent<PythonEnvsChangedEvent>,
+    ): AsyncIterator<PythonEnvInfo, void> {
+        const state = {
+            done: false,
+            pending: 0,
+        };
+        const seen: PythonEnvInfo[] = [];
+        const iterator = this.pythonEnvsManager.iterEnvs(query);
+
+        if (iterator.onUpdated !== undefined) {
+            iterator.onUpdated((event) => {
+                if (event === null) {
+                    state.done = true;
+                    checkIfFinishedAndNotify(state, didUpdate);
+                } else {
+                    const old = seen.find((s) => areSameEnvironment(s, event.old));
+                    if (old !== undefined) {
+                        state.pending += 1;
+                        resolveDifferencesInBackground(old, event.new, state, didUpdate, seen).ignoreErrors();
+                    }
+                }
+            });
+        }
+
+        let result = await iterator.next();
+        while (!result.done) {
+            const currEnv = result.value;
+            const old = seen.find((s) => areSameEnvironment(s, currEnv));
+            if (old !== undefined) {
+                state.pending += 1;
+                resolveDifferencesInBackground(old, currEnv, state, didUpdate, seen).ignoreErrors();
+            } else {
+                yield currEnv;
+                seen.push(currEnv);
+            }
+            // eslint-disable-next-line no-await-in-loop
+            result = await iterator.next();
+        }
+        if (iterator.onUpdated === undefined) {
+            state.done = true;
+            checkIfFinishedAndNotify(state, didUpdate);
+        }
+    }
+}
+
+async function resolveDifferencesInBackground(
+    oldEnv: PythonEnvInfo,
+    newEnv: PythonEnvInfo,
+    state: { done: boolean; pending: number },
+    didUpdate: EventEmitter<PythonEnvUpdatedEvent | null>,
+    seen: PythonEnvInfo[],
+) {
+    const merged = mergeEnvironments(oldEnv, newEnv);

```

Comment:
We don't want to emit an event if there is no change:

```suggestion
    const merged = mergeEnvironments(oldEnv, newEnv);
    if (deepEqual(merged, oldEnv) {
        return;
    }
```

or avoid `deepEqual()` by having `merge()` tell us:

```suggestion
    const merged = mergeEnvironments(oldEnv, newEnv);
    if (merged === undefined) {
        return;
    }
```

Reply:
It seems weird that `mergeEnvironments` returns `undefined` if nothing gets updated. In that case merge would just be the environment itself and I think that's fine to return.

So I'm going with the first one.